### PR TITLE
feat: support session affinity for AI providers

### PR DESCRIPTION
**Pull request by Aias00**
## Summary

- Extend `ai-load-balancer` `cluster_hash` to support session keys from headers, cookies, JSON request bodies, and Wasm metadata while preserving `hash_header` compatibility.
- Add `ai-proxy` provider-level `sessionAffinity` with `hash` and `persistent` modes, provider subset selection, optional scoping, and provider-unavailable policies (`failFast`, `fallbackAndUpdate`, `fallbackWithoutUpdate`).
- Add host-level runtime coverage for cluster hash routing and ai-proxy affinity selection/fallback behavior, plus Chinese and English docs.

## Test Plan

- [x] `go test ./...` in `plugins/wasm-go/extensions/ai-proxy`
- [x] `go test ./...` in `plugins/wasm-go/extensions/ai-load-balancer`
- [x] `go test . -run TestSessionAffinity -count=1` in `plugins/wasm-go/extensions/ai-proxy`
- [x] `git diff --check`

## Notes

- `fallbackAndUpdate` only updates persistent affinity mappings after a successful 2xx fallback response.
- Providers in the same `sessionAffinity.providerIds` set should use compatible protocols and API capabilities.

Related: #3840

### CODE REVIEW

File: plugins/wasm-go/extensions/ai-proxy/provider/retry.go [hunk 1]
```diff
@@ -59,6 +59,38 @@ func (c *ProviderConfig) retryFailedRequest(activeProvider Provider, ctx wrapper
...
+	c.setFallbackApiTokenInUse(ctx)
+	c.SetAvailableApiTokens(ctx)
+	modifiedHeaders, modifiedBody, err := c.transformRequestHeadersAndBody(ctx, activeProvider, [][2]string{
+		{"content-type", "application/json"},
+		{":authority", ctx.GetStringContext(CtxRequestHost, "")},
+		{":path", ctx.GetStringContext(CtxRequestPath, "")},
+	}, requestBody)
+	if err != nil {
+		return fmt.Errorf("fallback request failed to transform request headers and body: %v", err)
+	}
```
[the remark this answers is not in the document]
Comment (Aias00) [reply plugins/wasm-go/extensions/ai-proxy/provider/retry.go:74]: Handled in follow-up commit 524f1f99: the fallback transform error now uses %w so callers can unwrap the underlying error.

File: plugins/wasm-go/extensions/ai-load-balancer/cluster_hash/lb_policy.go [hunk 2]
```diff
@@ -86,26 +194,91 @@ func (lb ClusterHashLoadBalancer) selectCluster(hashKey string) string {
...
 func (lb ClusterHashLoadBalancer) HandleHttpRequestHeaders(ctx wrapper.HttpContext) types.Action {
-	hashKey, err := proxywasm.GetHttpRequestHeader(lb.HashHeader)
-	if err != nil || hashKey == "" {
-		log.Warnf("[ai-load-balancer/cluster_hash] missing hash header %q, rejecting request", lb.HashHeader)
-		_ = proxywasm.SendHttpResponse(403, nil, []byte("hash header required"), -1)
-		return types.ActionPause
+	if lb.Key.Source == hashKeySourceBody {
+		ctx.SetRequestBodyBufferLimit(lb.Key.MaxBodyBytes)
+		return types.HeaderStopIteration
+	}
```
[the remark this answers is not in the document]
Comment (Aias00) [reply plugins/wasm-go/extensions/ai-load-balancer/cluster_hash/lb_policy.go:203]: Handled in follow-up commit 524f1f99: body-key routing now rejects clearly bodyless requests (GET, HEAD, or content-length: 0) with 403 instead of unconditionally stopping iteration.

File: plugins/wasm-go/extensions/ai-proxy/config/config.go [hunk 3]
```diff
@@ -91,15 +203,544 @@ func (c *PluginConfig) Complete() error {
...
+	}
+	records[key] = sessionAffinityRecord{
+		ProviderID: providerID,
+		ExpiresAt:  now + c.TTLSeconds,
+	}
+	if err := c.storePersistentRecords(records, cas); err != nil {
+		return "", err
+	}
+	return providerID, nil
+}
```
[the remark this answers is not in the document]
Comment (Aias00) [reply plugins/wasm-go/extensions/ai-proxy/config/config.go:642]: Handled in follow-up commit 524f1f99: persistent session-affinity records are pruned before selection and update, with a regression test covering expired record cleanup.

**Review by wydream — COMMENTED** [review]
@Aias00 

Thanks for the contribution.
I would **request changes** before approving this PR.

The feature direction is reasonable and aligns with #3840. Backward compatibility also looks mostly preserved, and the relevant tests/CI are passing. However, I found a few correctness gaps that should be fixed before merge.

### Blocking / Should Fix

1. `cluster_hash` body-key routing can still stall bodyless requests  
   `plugins/wasm-go/extensions/ai-load-balancer/cluster_hash/lb_policy.go:196`

   When `key.source == body`, the code only rejects `GET`, `HEAD`, and `content-length: 0`. A `POST`/`PUT`/`PATCH` request with no body and no `content-length` can still return `HeaderStopIteration` and wait for a body callback that may never happen.  
   Please use the framework’s existing request-body detection logic or otherwise deterministically reject bodyless requests with 403. Add a runtime regression test for this case.

2. Persistent affinity can lose provider selection when shared-data write fails  
   `plugins/wasm-go/extensions/ai-proxy/config/config.go:619`

   `selectPersistentProviderID` computes a hash-selected provider, but if `storePersistentRecords` fails, it returns an error. The caller only logs the error and continues with the default active provider, which breaks affinity under CAS contention or transient shared-data failure.  
   Please keep routing to the selected provider and treat persistence as best-effort, or explicitly fail closed. Add test coverage for shared-data store/CAS failure.

3. Duplicate `sessionAffinity.providerIds` are not rejected or deduplicated  
   `plugins/wasm-go/extensions/ai-proxy/config/config.go:450`

   Duplicate IDs can enter `providerOrder`, and fallback traversal may select the same provider again instead of a real alternative.  
   Please reject duplicates or deduplicate them during validation, and add a validation test.

### Notes

- `hash_header` compatibility appears preserved for `ai-load-balancer`.
- `ai-proxy` currently fail-opens to the default provider when the session key is missing or extraction fails, while `ai-load-balancer` rejects missing hash keys with 403. This behavior difference should be documented or made explicit.
- Test coverage is good for happy paths and main fallback modes, but the edge cases above are currently uncovered.

**Review by wydream — CHANGES_REQUESTED** [review]
@Aias00 
Thanks for the update. I re-checked the previous review items and they look addressed:

- `cluster_hash` body-key routing now uses `ctx.HasRequestBody()` and has a regression test for bodyless POST requests.
- Persistent affinity now keeps routing to the hash-selected provider when shared-data persistence fails.
- Duplicate `sessionAffinity.providerIds` are now rejected with test coverage.

One remaining concern is about the semantics of `mode: persistent` in typical Higress deployments.

`ai-proxy` is usually deployed with multiple Higress/Envoy pods. The current persistent implementation stores mappings via Proxy-Wasm shared data (`GetSharedData` / `SetSharedData`), which is local to the current Envoy/Wasm instance. It is not backed by Redis or another external distributed store.

This means:

- affinity mappings are not shared across Higress pods;
- `fallbackAndUpdate` only updates the mapping in the pod that handled that request;
- mappings may be lost on pod restart, Wasm VM rebuild, or config reload;
- requests for the same session may still land on different providers if they are handled by different Higress replicas.

For stateless deterministic behavior across replicas, `mode: hash` is fine as long as all pods have the same provider order/configuration. But `mode: persistent` should probably be documented as local/best-effort persistence, not globally distributed persistence.

Could you please update the docs to explicitly call out this limitation and recommend `hash` mode for multi-replica deployments unless local best-effort remapping is acceptable? If global persistent affinity is required, it would need an external/shared storage design.

**Comment by Aias00** [comment]
> @Aias00 Thanks for the update. I re-checked the previous review items and they look addressed:
> 
> * `cluster_hash` body-key routing now uses `ctx.HasRequestBody()` and has a regression test for bodyless POST requests.
> * Persistent affinity now keeps routing to the hash-selected provider when shared-data persistence fails.
> * Duplicate `sessionAffinity.providerIds` are now rejected with test coverage.
> 
> One remaining concern is about the semantics of `mode: persistent` in typical Higress deployments.
> 
> `ai-proxy` is usually deployed with multiple Higress/Envoy pods. The current persistent implementation stores mappings via Proxy-Wasm shared data (`GetSharedData` / `SetSharedData`), which is local to the current Envoy/Wasm instance. It is not backed by Redis or another external distributed store.
> 
> This means:
> 
> * affinity mappings are not shared across Higress pods;
> * `fallbackAndUpdate` only updates the mapping in the pod that handled that request;
> * mappings may be lost on pod restart, Wasm VM rebuild, or config reload;
> * requests for the same session may still land on different providers if they are handled by different Higress replicas.
> 
> For stateless deterministic behavior across replicas, `mode: hash` is fine as long as all pods have the same provider order/configuration. But `mode: persistent` should probably be documented as local/best-effort persistence, not globally distributed persistence.
> 
> Could you please update the docs to explicitly call out this limitation and recommend `hash` mode for multi-replica deployments unless local best-effort remapping is acceptable? If global persistent affinity is required, it would need an external/shared storage design.

Updated the docs

**Review by wydream — APPROVED** [review]
LGTM

**Comment by Aias00** [comment]
### Implementation Rationale

This branch has been rebuilt on the current `main` (`16fe3f59052d1285d7dbf08bb707c35e9c2342d2`) while preserving the focused session-affinity implementation and its DCO-signed follow-up commits.

The overlapping `ai-proxy/main.go` change was resolved by keeping both behaviors:

- the current `main` request-size preflight and `content-length` handling;
- the session-affinity `allowBodyStop` gate required for body/model-based provider selection.

The final follow-up fixes a CAS retry bug in persistent affinity storage. A retry now reloads the latest shared-data map and reapplies only the intended key update, so it does not roll back concurrent updates to unrelated session keys. A regression test covers the stale `{A=oldA,B=oldB}` / latest `{A=newA,B=oldB}` / intended `B=newB` case.

Validation on exact head `055b4ae21f01818477c73ee9c02b0a01615ce777`:

- `go test ./config -run 'TestPluginConfig_SessionAffinity/persistent_store_cas_retry_preserves_concurrent_updates' -count=1`
- `go test ./config -count=1`
- `go test ./... -count=1` in `plugins/wasm-go/extensions/ai-proxy`
- `go test ./... -count=1` in `plugins/wasm-go/extensions/ai-load-balancer`
- `git diff --check origin/main...HEAD`
- independent exact-head review: zero P0/P1 findings

Scope remains limited to the approved session-affinity work related to #3840. No merge is requested or performed by this update.

**Comment by Aias00** [comment]
@wydream The branch has been rebuilt on the latest `main`, the conflict is resolved, and the current exact head is `055b4ae21f01818477c73ee9c02b0a01615ce777`. The focused and full plugin test suites listed above pass on this head. Please re-review when available.

File: plugins/wasm-go/extensions/ai-proxy/main.go [hunk 4]
```diff
@@ -218,7 +218,27 @@ func saveContextsToHeaders(ctx wrapper.HttpContext) {
...
+	}()
+
+	if pluginConfig.NeedsSessionAffinityRequestBody() && ctx.HasRequestBody() {
+		ctx.SetRequestBodyBufferLimit(pluginConfig.SessionAffinityMaxBodyBytes())
+		return types.HeaderStopIteration
+	}
+
+	if pluginConfig.IsSessionAffinityEnabled() && !pluginConfig.NeedsSessionAffinityRequestBody() {
+		if err := pluginConfig.SelectProviderBySessionAffinity(ctx, nil); err != nil {
+			log.Warnf("[onHttpRequestHeader] session affinity provider selection skipped: %v", err)
```
Comment (johnlanni) [inline plugins/wasm-go/extensions/ai-proxy/main.go:237]: [P2，非阻塞] session key 缺失/提取失败时这里仅 warn 并回退到默认（首个）provider，而 ai-load-balancer 对 body 来源缺 key 是 403 拒绝。两种语义都可接受，但建议在 README 中明确 ai-proxy 的"缺 key 回退默认 provider"行为，避免用户误以为请求会被拒绝。

File: plugins/wasm-go/extensions/ai-proxy/provider/retry.go [hunk 5]
```diff
@@ -60,6 +60,38 @@ func (c *ProviderConfig) retryFailedRequest(activeProvider Provider, ctx wrapper
 	return c.sendRetryRequest(ctx, apiName, activeProvider, retryClient, apiTokenInUse, apiTokens)
 }
 
+func (c *ProviderConfig) SendProviderFallbackRequest(ctx wrapper.HttpContext, activeProvider Provider, timeout uint32, callback func(int, http.Header, []byte)) error {
+	requestBody := ctx.GetByteSliceContext(CtxRequestBody, []byte(""))
```
Comment (johnlanni) [inline plugins/wasm-go/extensions/ai-proxy/provider/retry.go:64]: [P3，非阻塞] `CtxRequestBody` 仅在 provider 实现 RequestBodyHandler 且进入 body 处理时保存；对 original/passthrough 类 provider，fallback 请求可能携带空 body 导致 fallback provider 返回 4xx（随后恢复原响应，行为安全）。可考虑后续对这类场景补充说明或处理。

File: plugins/wasm-go/extensions/ai-proxy/config/config.go [hunk 6]
```diff
@@ -91,15 +208,561 @@ func (c *PluginConfig) Complete() error {
...
+		if len(c.Key.PropertyPath) == 0 {
+			if c.Key.Name == "" {
+				return fmt.Errorf("sessionAffinity.key.name or key.propertyPath is required when source is metadata")
+			}
+			c.Key.PropertyPath = []string{"metadata", c.Key.Name}
+		}
+	default:
+		return fmt.Errorf("unsupported sessionAffinity.key.source %q", c.Key.Source)
+	}
+	c.sharedDataKey = "ai-proxy-session-affinity:" + stableHash(strings.Join(c.providerOrder, ","))
```
Comment (johnlanni) [inline plugins/wasm-go/extensions/ai-proxy/config/config.go:507]: [P3，非阻塞] sharedDataKey 含 providerOrder 哈希：当 `providerIds` 集合变更时，旧 key 下的映射记录不会再被 load/prune，成为孤儿数据。由于单条记录小且仅在配置变更时产生，影响有限，可后续在文档或实现中补充清理策略。

**Review by johnlanni — DISMISSED** [review]
## 审查结论：APPROVE

绑定 exact head: `055b4ae21f01818477c73ee9c02b0a01615ce777`（base merge-base `16fe3f59`，三点 diff，20 个文件 +2627/-46）。

### 前置核对
- PR 状态: OPEN，非 Draft，headRefOid 与目标 head 一致，mergeable=MERGEABLE（无冲突，此前冲突已在 rebase 中解决）。
- 关联 Issue: 对应 feature 请求 #3840（OPEN，动机/配置示例/期望行为完整），PR 描述含 Summary/Test Plan/兼容性说明，信息完整性满足 feature 审查要求。
- CI: lint、build、coverage-test、CodeQL、higress-conformance-test、gateway-conformance-test (v1.4.0/v1.6.0)、inference-extension-conformance-test、higress-wasmplugin-test (GO/RUST)、Test Changed Plugins (ai-proxy / ai-load-balancer)、Test Summary & Coverage、DCO 全部 SUCCESS；MCP 2026 Conformance 为条件性 SKIPPED（本 PR 未涉及 MCP 插件）；Dependency Quality / License Compliance / Security Analysis / yy-sensitive-scan 四个 commit status 均为 success。

### 代码审查要点
- `ai-load-balancer/cluster_hash`: 新增 header/cookie/body/metadata 四种 key 来源，保持 `hash_header` 向后兼容；body 来源通过 `ctx.HasRequestBody()` + GET/HEAD 检查拒绝无 body 请求（403），修复了此前 Copilot 指出的 HeaderStopIteration 挂起问题；`max_body_bytes` 有 uint32 上界校验。
- `ai-proxy/config`: sessionAffinity 支持 hash/persistent 两种模式、providerIds 子集（重复/未知 id 均校验拒绝）、scope 组合、failFast/fallbackAndUpdate/fallbackWithoutUpdate 三种不可用策略；persistent 映射存于 Wasm shared data，读取时裁剪过期记录，CAS 冲突重试会重新加载最新 map 并只重放当前 key 的更新（055b4ae 修复并发覆盖），共享数据失败时退化为 hash 选择且不阻塞请求；sharedDataKey 含 providerOrder 哈希避免不同 provider 集合串读。
- `ai-proxy/main.go`: body/metadata 取 key 时先缓冲请求体再选 provider（`allowBodyStop` 门控避免二次停顿），选择失败时告警并回退默认 provider；响应阶段按选中 provider 处理，`unavailableStatus`（默认 5.*）命中后按策略 failFast(503) 或 fallback（仅 2xx 时更新 persistent 映射，非 2xx 恢复原响应）。
- 安全/隐私: 日志不再输出 API token（此前 Copilot 意见已修复，`setFallbackApiTokenInUse` 仅打固定文案）；key 值本身不落日志；错误包装改用 `%w`。
- 文档: 中英文 README 均说明 persistent 为本地尽力而为持久化、多副本建议 hash 模式、协议兼容性约束，回应了 wydream 此前关于跨副本语义的意见。
- 先前所有 review 意见（Copilot ×3、wydream）在本 head 中均已处理。

### 本地定向测试（exact head，带 timeout）
- `go test ./cluster_hash/... . -count=1`（ai-load-balancer）: ok（9.9s，含 go+wasm 运行时用例）
- `go test ./config -count=1`（ai-proxy）: ok（含 persistent CAS 并发保留回归测试）
- `go test . -run TestSessionAffinity -count=1`（ai-proxy）: PASS（header/cookie/metadata 选择、failFast、fallbackAndUpdate 重映射、5xx fallback，go 与 wasm 双模式共 10 子用例）
- `go vet` 变更包: 仅存量的 provider 结构体 json tag 告警，无新增问题
- 全仓 e2e/单测由 CI 覆盖并全部 SUCCESS

### 非阻塞建议（详见行级评论，均为 P2/P3）
1. ai-proxy 中 session key 缺失时静默回退默认 provider（warn 日志），与 ai-load-balancer body 来源 403 的语义不一致，建议在 README 中补充说明；
2. fallback 请求体依赖 `CtxRequestBody`，仅当 provider 实现 RequestBodyHandler 时保存，passthrough 场景可能发出空 body fallback；
3. providerIds 集合变更后旧 shared-data key 下的记录成为孤儿数据，量小且有 TTL 裁剪兜底，可后续优化。

未发现 P0/P1 问题，无安全、并发或兼容性 blocker，CI 全绿，予以 APPROVE。

File: plugins/wasm-go/extensions/ai-proxy/provider/retry.go (hunk 5 above)
Comment (Aias00) [reply plugins/wasm-go/extensions/ai-proxy/provider/retry.go:64]: 确认。当前 fallback 是 best effort：未保存 `CtxRequestBody` 时 fallback 可能失败，回调收到非 2xx 后会恢复原响应，因此不会替换主请求结果。让 passthrough/original provider 始终缓存并重放 body 会扩大请求体生命周期和内存边界，需要独立设计与回归，本 PR 不扩展该范围。

File: plugins/wasm-go/extensions/ai-proxy/config/config.go (hunk 6 above)
Comment (Aias00) [reply plugins/wasm-go/extensions/ai-proxy/config/config.go:507]: 确认。`providerIds`/order 变化后会切换 shared-data key，旧 key 在当前 Wasm VM 生命周期内可能残留，且不会再被新配置主动 prune；它只在 provider 集合变更时产生，影响有限。本 PR 保持不同 provider 集合的数据隔离，清理或迁移策略后续单独处理。

File: plugins/wasm-go/extensions/ai-proxy/main.go (hunk 4 above)
Comment (Aias00) [reply plugins/wasm-go/extensions/ai-proxy/main.go:237]: 收到。这里保留 ai-proxy 的容错语义：session key 缺失或提取失败时记录 warn 并使用默认 provider，不返回 403；ai-load-balancer 的 403 仅适用于 cluster_hash body key 无法取得的场景。该差异作为文档跟进，不在已经通过 exact-head 审查的本 PR 中追加改动。

**Review by johnlanni — CHANGES_REQUESTED** [review]
@wydream 该 PR 暂不合并。

此前的 APPROVE 已撤回；本条是当前的合入结论，替代此前的合入信号。

原因不是本轮已发现的代码正确性问题，而是当前设计要求在 `providers` 下配置多个可选 provider。这个方案的配置模型、回退语义以及用户和运维侧的复杂度都比较高，需要 @wydream 后续进一步评估：是否应以这种方式暴露多 provider 选择、如何控制配置复杂度，以及后续演进边界。

在评估结论明确前，请不要合并。若后续确定继续推进或调整设计，请在新的 exact head 上重新 review。

本条为非代码正确性阻塞的设计暂缓说明。

**Comment by Aias00** [comment]
Rebased #3921 onto current `main`; the new head is `e9a3dddc`.

Conflict resolution preserved the reviewed session-affinity implementation and current main changes. The only follow-up commit updates the `cluster_metrics` test fake for the newer `wrapper.HttpContext` body-inspection methods required by this PR's wasm-go version; it has no production behavior change.

Validated locally:
- `go test ./... -count=1` in `plugins/wasm-go/extensions/ai-load-balancer`
- `go test ./... -count=1` in `plugins/wasm-go/extensions/ai-proxy`
- `git diff --check higress-group/main...HEAD`

The existing design hold remains unchanged; this update only resolves integration conflicts and refreshes CI. @wydream @johnlanni please re-evaluate on the new head when the design decision is ready.
